jax/experimental/custom_partitioning.py

```diff
@@ -130,39 +130,190 @@ def _default_propagate_user_shardings(sharding, shape):
+      param = c64[16384,1024]{1,0} parameter(0), sharding={devices=[2,1]0,1}
+      all-gather = c64[32768,1024]{1,0} all-gather(param), channel_id=1, replica_groups={{0,1}}, dimensions={0}, use_global_device_ids=true, metadata={op_name="pjit(fft)/jit(main)/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="doc.py" source_line=42}
+      fft.1 = c64[32768,1024]{1,0} fft(all-gather), fft_type=FFT, fft_length={1024}, metadata={op_name="pjit(fft)/jit(main)/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="doc.py" source_line=42}
+      partition-id = u32[] partition-id(), metadata={op_name="pjit(fft)/jit(main)/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="doc.py" source_line=42}
+      ROOT fusion = c64[16384,1024]{1,0} fusion(fft.1, partition-id), kind=kLoop, calls=fused_computation, metadata={op_name="pjit(fft)/jit(main)/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="doc.py" source_line=42}
+    }
+
+  Where the ``all-gather`` before the FFT and the dynamic-slice after are both clearly visible.
+  This means that the input is gathered on all devices prior to the FFT, and sliced after.
+
+  ``pjit_my_fft``, on the other hand, simply compiles to
+
+  .. code-block::
+
+    HloModule pjit__unnamed_wrapped_function_, entry_computation_layout={(c64[16384,1024]{1,0})->c64[16384,1024]{1,0}}
+
+    ENTRY main.5_spmd {
+      param = c64[16384,1024]{1,0} parameter(0), sharding={devices=[2,1]0,1}
+      ROOT fft.0 = c64[16384,1024]{1,0} fft(param), fft_type=FFT, fft_length={1024}, metadata={op_name="jit(main)/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="doc.py" source_line=41}
+    }
+
+  where no unnecessary sharding is taking place.
+
+  Because of the logic in ``supported_sharding``, ``my_fft`` also works on 1-dimensional arrays.
+
+  >>> with Mesh(np.array(jax.devices()), ('x',)):
+  ...   x = np.asarray(np.random.randn(32*1024*1024), dtype=np.complex64)
+  ...   y = pjit(lambda x: x, in_axis_resources=None, out_axis_resources=P('x'))(x)
+  ...   pjit_my_fft = pjit(my_fft, in_axis_resources=P('x'), out_axis_resources=P('x'))
+  ...   pjit_fft    = pjit(fft,    in_axis_resources=P('x'), out_axis_resources=P('x'))
+  ...   print(pjit_my_fft(y))
+  ...   print(pjit_fft(y))
+
+  .. code-block::
+
+    # my_fft
+    [    7.217285   +0.j     -3012.4937  +4287.635j   -405.83594 +3042.984j
+    ...  1422.4502  +7271.4297j  -405.84033 -3042.983j
+    -3012.4963  -4287.6343j]
+
+    # jax.numpy.fft.fft
+    [    7.217285   +0.j     -3012.4937  +4287.635j   -405.83594 +3042.984j
+    ...  1422.4502  +7271.4297j  -405.84033 -3042.983j
+    -3012.4963  -4287.6343j]
+
+  In this case, the HLO of ``my_fft`` does show an all-gather and dynamic-slice, since the last dimension
+  is the dimension along which FFTs are calculated.
+
+  .. code-block::
+
+    HloModule pjit__unnamed_wrapped_function_, entry_computation_layout={(c64[16777216]{0})->c64[16777216]{0}}
```

Comment:
Can we remove the entire HLO print here since it can get stale?

You can just do an assert in code to test for dynamic-slice, etc

Reply:
Can you point me to an example or the right API to use to do that?

Reply:
I think yash is suggesting just to do re.match on the output_text for dynamic-slice.

Reply:
Pushed a new version ; please have a look, thanks!

Reply:
Looks like there are still some presubmits failing.

Reply:
Can https://github.com/google/jax/actions/runs/3991188021/jobs/6845708206 be skipped for a particular file/section/... somehow?

Otherwise, I guess I need to put everything in a single `code-block`, which is fine, but seems a little less-nice for this specific case.

I'll fix the other two.

Reply:
I think the problem is not that they're two codeblocks its just that pytest is picking up on the

>>> code()
output

form but ignoring the code-block form.

doctests seem to carry over from one call to the next if you use the >>> form for all of them. I dunno how the docs look in that case.

Reply:
Fixed (WARed, really) using code-blocks. I thinks that should do it.

Reply:
All precommit checks are passing now